server/src/main/java/com/flipkart/varadhi/web/AuthnHandler.java

```diff
@@ -1,108 +1,128 @@
+                throw new InvalidConfigException(
+                    "Provider class " + providerClass.getName()
+                                                 + " does not implement AuthenticationHandlerProvider interface"
+                );
             }
-            jwtAuthOptions.setJWTOptions(config.getOptions());
-            JWTAuth provider = JWTAuth.create(vertx, jwtAuthOptions);
-            return JWTAuthHandler.create(provider);
+            provider = (AuthenticationHandlerProvider)providerClass.getDeclaredConstructor().newInstance();
+
+
+        } catch (ClassNotFoundException e) {
+            throw new InvalidConfigException(
+                "Authentication handler provider class not found: " + authenticationConfig
+                                                                                          .getHandlerProviderClassName(),
+                e
+            );
         } catch (Exception e) {
-            throw new VaradhiException("Failed to Initialise JWT Authentication handler.", e);
+            throw new InvalidConfigException("Failed to create authentication handler provider", e);
+        }
+
+        try {
+            authenticationHandler = new AuthenticationHandlerWrapper(
+                provider.provideHandler(vertx, JsonObject.mapFrom(authenticationConfig), Org::of, meterRegistry),
+                authenticationConfig.getWhitelistedURLs()
+            );
+        } catch (Exception e) {
+            throw new InvalidConfigException("Failed to create authentication handler", e);
         }
     }
 
-    AuthenticationHandler createUserHeaderHandler() {
-        return SimpleAuthenticationHandler.create().authenticate(ctx -> {
-            String userName = ctx.request().getHeader(USER_ID_HEADER);
-            if (StringUtils.isBlank(userName)) {
-                return Future.failedFuture(new HttpException(HTTP_UNAUTHORIZED, "no user details present"));
-            }
-            return Future.succeededFuture(User.fromName(userName));
-        });
+    public void configure(Route route, RouteDefinition routeDef) {
+        if (authenticationHandler != null) {
+            route.handler(authenticationHandler);
+        }
     }
 
-    @AllArgsConstructor
     static class AuthenticationHandlerWrapper implements Handler<RoutingContext> {
         private final Handler<RoutingContext> wrappedHandler;
+        private final URLMatcherUtil whitelistedURLMatcher;
+
+        public AuthenticationHandlerWrapper(
+            Handler<RoutingContext> wrappedHandler,
+            List<URLDefinition> whitelistedURLs
+        ) {
+            this.wrappedHandler = wrappedHandler;
+            this.whitelistedURLMatcher = new URLMatcherUtil(whitelistedURLs);
+        }
 
         @Override
         public void handle(RoutingContext ctx) {
-            wrappedHandler.handle(ctx);
+            if (this.whitelistedURLMatcher.matches(String.valueOf(ctx.request().method()), ctx.request().path())) {
```

Comment:
NIT: THere is a possibility of moving the whitelisting / orgContextExemption regex pattern checks to be moved to app startup, when the handlers are initialized. All to save the pattern match cost for every request.